src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,426 @@
+                    },
+                }
+            };
+            let current_user_key = Key::from_encoded(current_user_key);
+
+            // Attempt to read specified version of the key. Note that we may get `None`
+            // indicating that no desired version is found, or a DELETE version is found, or
+            // when `has_write == false`. We may also get `Err` due to key lock. We need to
+            // ensure that the scanner keeps working for future calls even when meeting key
+            // lock errors (why??), so we don't apply `?` operator to the result here thus
+            // iterators can be moved forward.
+            let result = self.get(&current_user_key, has_write, has_lock);
+
+            if has_write {
+                self.move_write_cursor_to_next_user_key(&current_user_key)?;
+            }
+            if has_lock {
+                self.lock_cursor.next(&mut self.statistics.lock);
+            }
+
+            // If we got something, it can be just used as the return value. Otherwise, we need
+            // to continue stepping the cursor.
+            if let Some(v) = result? {
+                return Ok(Some((current_user_key, v)));
+            }
+        }
+    }
+
+    /// Attempt to get the value of a key specified by `user_key` and `self.ts`. This function
+    /// requires that the write cursor is currently pointing to the latest version of `user_key`.
+    #[inline]
+    fn get(&mut self, user_key: &Key, has_write: bool, has_lock: bool) -> Result<Option<Value>> {
+        let mut safe_ts = self.ts;
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                if has_lock {
+                    assert!(self.lock_cursor.valid());
+                    let lock_value = self.lock_cursor.value(&mut self.statistics.lock);
+                    let lock = Lock::parse(lock_value)?;
+                    safe_ts = super::util::check_lock(user_key, safe_ts, &lock)?
+                }
+            }
+            IsolationLevel::RC => {}
+        }
+
+        if !has_write {
+            return Ok(None);
+        }
+        assert!(self.write_cursor.valid());
+
+        // The logic starting from here is similar to `PointGetter`.
+
+        // Try to iterate to `${user_key}_${safe_ts}`. We first `next()` for a few times,
+        // and if we have not reached where we want, we use `seek()`.
+
+        // Whether we have *not* reached where we want by `next()`.
+        let mut needs_seek = true;
+
+        for i in 0..SEEK_BOUND {
+            if i > 0 {
+                self.write_cursor.next(&mut self.statistics.write);
+                if !self.write_cursor.valid() {
+                    // Key space ended.
+                    return Ok(None);
+                }
+            }
+            {
+                let current_key = self.write_cursor.key(&mut self.statistics.write);
+                if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice()) {
```

Comment:
We can use this information to skip `move_write_cursor_to_next_user_key` in `read_next`.

Reply:
Yes. I've added TODO here as well.